run/O2SimDeviceRunner.cxx

```diff
@@ -35,20 +35,19 @@ namespace bpo = boost::program_options;
 std::vector<int> gChildProcesses; // global vector of child pids
 int gMasterProcess = -1;
 
-// custom signal handler to ensure that we
-// distribute signals to all detached forks
-void sighandler(int signal)
+void sigaction_handler(int signal, siginfo_t* signal_info, void*)
 {
-  if (signal == SIGTERM || signal == SIGINT) {
-    auto pid = getpid();
-    if (pid == gMasterProcess) {
-      // the master
-      for (auto child : gChildProcesses) {
-        kill(child, signal);
-      }
-    }
-    _exit(0);
+  auto pid = getpid();
+  LOG(INFO) << pid << " caught signal from source " << signal_info->si_pid;
+  auto groupid = getpgrp();
+  if (pid == gMasterProcess) {
+    killpg(pid, signal); // master kills whole process group
+  } else {
+    // forward to master
+    kill(groupid, signal);
   }
+  _exit(1); // we treat signal interruption as an error
```

Comment:
I am not 100% sure, but are we sure the only signal we will receive is KILL and TERM? Perhaps we should have a filter for the singal in here?

Reply:
Why? The filter is specified further below in the main function.

Reply:
ok, sorry, I missed that. Then it is OK. Perhaps we should also forward `SIGQUIT` and `SIGKILL`? Or probably SIGKILL cannot be handled?

Reply:
SIGKILL can't be handled.

Reply:
ok, I would put that in a comment to avoid confusion, and perhaps I'd add SIGQUIT just in case it is used at some time to avoid misleading behavior.